### PR TITLE
feat(es/lint): no-alert implementation

### PR DESCRIPTION
**Pull request by ArturAralin**
**Description:**

no-alert rule implementations.

Note: current implementation is not supports calls on "this" like `this.alert('msg')`

**BREAKING CHANGE:**
No

### CODE REVIEW

File: crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr
Comment (kdy1) [inline crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: I think this should not be an error, as user is calling user-defined `alert`.
If we want to report an error, it should be usage before declarationn.

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr
Comment (ArturAralin) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: you're mean is should be caught by another rule?

Comment (kdy1) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: I meant this should not be treated as a call to the global `alert`.

Comment (ArturAralin) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: Ok. I'll fix it

Comment (ArturAralin) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: @kdy1 while I investigating how to solve your comment I did found different syntax context resolving behaviour in tests in swc and swc_ecma_lint crates. I managed to solve different behaviour by adding to `crates/swc/src/config/mod.rs` following command 

```
program.visit_mut_with(&mut resolver_with_mark(top_level_mark));
```

It is correct solution or it should be fixed with another way?

Comment (kdy1) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: It's wrong and I think we need a feature gate for non-critical errors.

Comment (ArturAralin) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: So I need help with understanding how to fix problem with different syntax contexts resolving 😃

Comment (kdy1) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: Ah, are you using collecting bindings on creating?

Comment (kdy1) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: I think you can do it in `lint_module` instead?
If it does not work, it needs some thinking about desgin...

Comment (kdy1) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: Oh, I think your fix is correct way to go (applying resolver), although we need to remove the next call to the resolver.

Comment (ArturAralin) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: Okay. Happy to hear it

Comment (ArturAralin) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: hm... My fix solves my problem, but it fail some tests 

```
    tests_tests__fixture__issue_1160__input
    tests_tests__fixture__issue_1345__input
    tests_tests__fixture__issue_1362__case1__input
    tests_tests__fixture__issue_1421__case1__input
    tests_tests__fixture__issue_1869__decorator__input
    tests_tests__fixture__issue_846__case1__input
 ```

Comment (ArturAralin) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: I can commit my fixes or provide terminal output

Comment (kdy1) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: I'll create another PR to fix the wrong design and you can rebase after it's merged.

Comment (kdy1) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: I made https://github.com/swc-project/swc/pull/3400

Comment (ArturAralin) [reply crates/swc/tests/errors/lints/no-alert/alert-ident-hoisting/output.swc-stderr:4]: Thank you

**Comment by ArturAralin** [comment]
@kdy1 what about feature gate, I'd make it later

**Comment by kdy1** [comment]
It's to prevent binary size regression of `nexg-swc` and other users of `swc` crate.
I think `no-alert` and `no-console` should be removed by the feature gate by default, and enabled only for `@swc/core` or `@swc/wasm*`

**Comment by ArturAralin** [comment]
Ok. So if I'd add feature gate in https://github.com/swc-project/swc/pull/3399 it's okay? Just don't want update two other opened PRs

**Comment by kdy1** [comment]
Okay, I think it would be fine.

**Comment by ArturAralin** [comment]
@kdy1 good. Then I'll be wait for you review this PR

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/no_alert.rs:57]: Why is this required?

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/no_alert.rs:58]: Same here, we can just use `static` item from rule.

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/no_alert.rs:133]: This impl seems redundant.
This is default implementation of `visit_opt_chain_expr`.

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/no_alert.rs:127]: I think this will be broken for nested calls.

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_alert.rs:127]: e.g.

```js
alert(() => alert("foo"))
```

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/no_alert.rs:137]: I think handling `Expr::Ident` from `visit_expr` is generally better, as it's much more efficient and much likely to work properly.

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_alert.rs:137]:
```js
(() => {
    const obj = {
        alert,
    };
})()
```

It's in a call, so `call_expr_span` is `Some` and `visit_ident` is called also for shorthand properties.

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_alert.rs:137]: Also,

```js
(() => {
    alert = () => {};
})()
```

it will have the top-level context and `visit_ident` is also called for patterns.

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_alert.rs:137]: One more,

```js
(() => {
    console.log(alert);
})()
```

is this a target of the lint rule?

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/no_alert.rs:58]: You're mean that I can use static hashset?

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/no_alert.rs:127]: Yes, you're right

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_alert.rs:58]: No, you don't need hastset for this.

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/no_alert.rs:139]: I think this will error for

```js
foo(() => alert("foo"));
```

as `call_expr_span` points `foo` call while vaildating `alert`.

File: crates/swc_ecma_lints/src/rules/mod.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/mod.rs:9]: Can you remove `pub`?

File: crates/swc_ecma_lints/src/rules/mod.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/mod.rs:10]: Also, I missed this one on previous review. Can you remove `pub` here, too?

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/rules/no_alert.rs:97]: Using a nested visitor without a stopper method is bad for performance.

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_alert.rs:97]: Also, we don't need two visitor here, and visitor is quite big in final binary.

Comment (kdy1) [reply crates/swc_ecma_lints/src/rules/no_alert.rs:97]: (That's why methods for collecting bindings uses single visitor)

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc_ecma_lints/src/rules/no_alert.rs
Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/no_alert.rs:97]: Hmm.. Seems like I need advice how visit indents/members only on callee

Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/no_alert.rs]: I rewrote implementation without nested visitor, but it looks a bit awkward by my mind

File: crates/swc_ecma_lints/src/rules/mod.rs
Comment (ArturAralin) [reply crates/swc_ecma_lints/src/rules/mod.rs]: changed in on pub(crate). Can't remove `pub` because config structs placed into modules

**Comment by ArturAralin** [comment]
@kdy1 also I did feature gate

File: crates/swc_ecma_lints/src/config.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/config.rs:53]: This will make these lints removed when the featrure is enabled.
Can you revert commits related to feature gates?

Comment (ArturAralin) [reply crates/swc_ecma_lints/src/config.rs:53]: A bit didn't understand you. Why lints will be removed if features is enabled? Here is `cfg(not(feature...`

Comment (kdy1) [reply crates/swc_ecma_lints/src/config.rs:53]: Because `#[cfg(not(feature = "non_critical_lints"))]` means this code (under `cfg`) should be built only if `not(feature = "non_critical_lints")`.
It means if we enable `non_critical_lints `, it's `not(true)`, so it's removed.

Comment (ArturAralin) [reply crates/swc_ecma_lints/src/config.rs:53]: then if we're not enable non_critical_lints then it's not(false) -> true.

if it is then it's expected.

here is I include all non critical lints if feature flag provided https://github.com/swc-project/swc/pull/3394/files/aade1992b77528596ce47ab63fa45ecf2e4374e3#diff-804fa350bc6183d7f8eb97b783ba6090005850a1f7a3c176f62b572871ff558dR1

and here is not https://github.com/swc-project/swc/pull/3394/files/aade1992b77528596ce47ab63fa45ecf2e4374e3#diff-804fa350bc6183d7f8eb97b783ba6090005850a1f7a3c176f62b572871ff558dR53

Comment (ArturAralin) [reply crates/swc_ecma_lints/src/config.rs:53]: Here I provide empty LintConfig struct if `non_critical_lints` feature flag isn't provided

Comment (ArturAralin) [reply crates/swc_ecma_lints/src/config.rs:53]: Possibly here is should be used another name

https://github.com/swc-project/swc/pull/3394/files/aade1992b77528596ce47ab63fa45ecf2e4374e3#diff-804fa350bc6183d7f8eb97b783ba6090005850a1f7a3c176f62b572871ff558dR54

File: crates/swc/Cargo.toml
Comment (kdy1) [inline crates/swc/Cargo.toml:53]: This should not be enabled by the swc crate.
`node` and `wasm` should depend on `swc_ecma_lints` and enable this feature.

File: crates/swc_ecma_lints/src/config.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/config.rs:2]: I think this can be removed?

File: crates/swc_ecma_lints/src/config.rs
Comment (kdy1) [inline crates/swc_ecma_lints/src/config.rs:10]: I think this type should exist regardless of features.
The only things we need to feature-gate is impl itself (module) and the function `all`.

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc/Cargo.toml
Comment (ArturAralin) [reply crates/swc/Cargo.toml:53]: Then tests in `swc` crate will fail

Comment (kdy1) [reply crates/swc/Cargo.toml:53]: Then we can add `swc_ecma_lints ` with features enabled to `[devDependencies]`, as we use 2021 edition.

Comment (ArturAralin) [reply crates/swc/Cargo.toml:53]: Sounds good

File: crates/swc/Cargo.toml
Comment (kdy1) [inline crates/swc/Cargo.toml:94]: Please add version here, otherwise `cargo` will deny publishing.

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you! Code changes look good to me.

File: crates/swc/Cargo.toml
Comment (ArturAralin) [reply crates/swc/Cargo.toml:94]: oops. missed it

**Review by kdy1 — APPROVED** [review]
Thank you!

**Pull request merged**
